File: agentos/api/routers/webhooks.py
"""Webhooks router — CRUD + test delivery + deliveries + secret rotation."""

from __future__ import annotations

import hashlib
import hmac
import ipaddress
import json
import secrets
import time
import uuid
from urllib.parse import urlparse

from fastapi import APIRouter, Depends, HTTPException

from agentos.api.deps import CurrentUser, get_current_user, _get_db
from agentos.api.schemas import CreateWebhookRequest, WebhookResponse

router = APIRouter(prefix="/webhooks", tags=["webhooks"])


def _validate_callback_url(url: str) -> None:
    parsed = urlparse(url)
    host = parsed.hostname or ""
    if parsed.scheme not in {"http", "https"} or not host:
        raise HTTPException(status_code=400, detail="Invalid webhook URL")
    if host in {"localhost"} or host.endswith(".local") or host.endswith(".internal"):
        raise HTTPException(status_code=400, detail="Webhook URL host is not allowed")
    try:
        ip = ipaddress.ip_address(host)
        if ip.is_private or ip.is_loopback or ip.is_link_local or ip.is_reserved:
            raise HTTPException(status_code=400, detail="Webhook URL host is not allowed")
    except ValueError:
        # Non-IP hostnames are allowed; DNS-level filtering can be added upstream.
        pass


@router.get("", response_model=list[WebhookResponse])
async def list_webhooks(user: CurrentUser = Depends(get_current_user)):
    db = _get_db()
    rows = db.conn.execute(
        "SELECT * FROM webhooks WHERE org_id = ? ORDER BY created_at DESC", (user.org_id,)
    ).fetchall()
    return [
        WebhookResponse(
            webhook_id=r["webhook_id"], url=r["url"],
            events=json.loads(r["events"]), is_active=bool(r["is_active"]),
            failure_count=r["failure_count"], last_triggered_at=r["last_triggered_at"],
        )
        for r in rows
    ]


@router.post("", response_model=WebhookResponse)
async def create_webhook(request: CreateWebhookRequest, user: CurrentUser = Depends(get_current_user)):
    url_str = str(request.url)
    _validate_callback_url(url_str)
    db = _get_db()
    webhook_id = uuid.uuid4().hex[:12]
    secret = uuid.uuid4().hex

    db.conn.execute(
        "INSERT INTO webhooks (webhook_id, org_id, url, secret, events) VALUES (?, ?, ?, ?, ?)",
        (webhook_id, user.org_id, url_str, secret, json.dumps(request.events)),
    )
    db.conn.commit()

    return WebhookResponse(
        webhook_id=webhook_id, url=url_str, events=request.events,
    )


@router.put("/{webhook_id}")
async def update_webhook(
    webhook_id: str,
    url: str = "",
    events: list[str] | None = None,
    is_active: bool | None = None,
    user: CurrentUser = Depends(get_current_user),
):
    """Update a webhook."""
    db = _get_db()
    updates, params = [], []
    if url:
        _validate_callback_url(url)
        updates.append("url = ?")
        params.append(url)
    if events is not None:
        updates.append("events = ?")
        params.append(json.dumps(events))
    if is_active is not None:
        updates.append("is_active = ?")
        params.append(1 if is_active else 0)
    if not updates:
        raise HTTPException(status_code=400, detail="Nothing to update")
    params.extend([webhook_id, user.org_id])
    db.conn.execute(f"UPDATE webhooks SET {', '.join(updates)} WHERE webhook_id = ? AND org_id = ?", params)
    db.conn.commit()
    return {"updated": webhook_id}


@router.delete("/{webhook_id}")
async def delete_webhook(webhook_id: str, user: CurrentUser = Depends(get_current_user)):
    db = _get_db()
    result = db.conn.execute(
        "DELETE FROM webhooks WHERE webhook_id = ? AND org_id = ?", (webhook_id, user.org_id)
    )
    db.conn.commit()
    if result.rowcount == 0:
        raise HTTPException(status_code=404, detail="Webhook not found")
    return {"deleted": webhook_id}


@router.post("/{webhook_id}/test")
async def test_webhook(webhook_id: str, user: CurrentUser = Depends(get_current_user)):
    """Send a test event to a webhook."""
    import httpx

    db = _get_db()
    row = db.conn.execute(
        "SELECT * FROM webhooks WHERE webhook_id = ? AND org_id = ?", (webhook_id, user.org_id)
    ).fetchone()
    if not row:
        raise HTTPException(status_code=404, detail="Webhook not found")

    webhook = dict(row)
    payload = {
        "event": "test",
        "timestamp": time.time(),
        "data": {"message": "This is a test webhook delivery from AgentOS"},
    }

    try:
        start = time.monotonic()
        async with httpx.AsyncClient(timeout=10) as client:
            resp = await client.post(
                webhook["url"],
                json=payload,
                headers={"X-AgentOS-Secret": webhook["secret"]},
            )
        duration = (time.monotonic() - start) * 1000

        # Log delivery
        db.conn.execute(
            """INSERT INTO webhook_deliveries (webhook_id, event_type, payload_json,
            response_status, response_body, duration_ms, success) VALUES (?, ?, ?, ?, ?, ?, ?)""",
            (webhook_id, "test", json.dumps(payload), resp.status_code,
             resp.text[:500], duration, 1 if resp.status_code < 400 else 0),
        )
        db.conn.commit()

        return {"status": resp.status_code, "duration_ms": round(duration, 1), "success": resp.status_code < 400}
    except Exception as exc:
        return {"status": 0, "error": str(exc), "success": False}


@router.get("/{webhook_id}/deliveries")
async def list_deliveries(
    webhook_id: str,
    limit: int = 50,
    user: CurrentUser = Depends(get_current_user),
):
    """List delivery attempts for a webhook with status codes."""
    db = _get_db()
    # Verify webhook belongs to user's org
    wh = db.conn.execute(
        "SELECT webhook_id FROM webhooks WHERE webhook_id = ? AND org_id = ?",
        (webhook_id, user.org_id),
    ).fetchone()
    if not wh:
        raise HTTPException(status_code=404, detail="Webhook not found")

    rows = db.conn.execute(
        "SELECT * FROM webhook_deliveries WHERE webhook_id = ? ORDER BY rowid DESC LIMIT ?",
        (webhook_id, limit),
    ).fetchall()
    return {"deliveries": [dict(r) for r in rows]}


@router.post("/{webhook_id}/deliveries/{delivery_id}/replay")
async def replay_delivery(
    webhook_id: str,
    delivery_id: str,
    user: CurrentUser = Depends(get_current_user),
):
    """Replay a failed webhook delivery."""
    import httpx

    db = _get_db()
    # Verify webhook ownership
    wh = db.conn.execute(
        "SELECT * FROM webhooks WHERE webhook_id = ? AND org_id = ?",
        (webhook_id, user.org_id),
    ).fetchone()
    if not wh:
        raise HTTPException(status_code=404, detail="Webhook not found")
    webhook = dict(wh)

    # Find original delivery
    delivery = db.conn.execute(
        "SELECT * FROM webhook_deliveries WHERE rowid = ? AND webhook_id = ?",
        (delivery_id, webhook_id),
    ).fetchone()
    if not delivery:
        raise HTTPException(status_code=404, detail="Delivery not found")
    delivery = dict(delivery)

    payload = json.loads(delivery.get("payload_json", "{}"))

    try:
        start = time.monotonic()
        async with httpx.AsyncClient(timeout=10) as client:
            resp = await client.post(
                webhook["url"],
                json=payload,
                headers={"X-AgentOS-Secret": webhook["secret"]},
            )
        duration = (time.monotonic() - start) * 1000

        # Log the replay delivery
        db.conn.execute(
            """INSERT INTO webhook_deliveries (webhook_id, event_type, payload_json,
            response_status, response_body, duration_ms, success) VALUES (?, ?, ?, ?, ?, ?, ?)""",
            (webhook_id, delivery.get("event_type", "replay"), json.dumps(payload),
             resp.status_code, resp.text[:500], duration, 1 if resp.status_code < 400 else 0),
        )
        db.conn.commit()

        return {
            "replayed": delivery_id,
            "status": resp.status_code,
            "duration_ms": round(duration, 1),
            "success": resp.status_code < 400,
        }
    except Exception as exc:
        return {"replayed": delivery_id, "status": 0, "error": str(exc), "success": False}


@router.post("/{webhook_id}/rotate-secret")
async def rotate_secret(
    webhook_id: str,
    user: CurrentUser = Depends(get_current_user),
):
    """Rotate the signing secret for a webhook."""
    db = _get_db()
    row = db.conn.execute(
        "SELECT webhook_id FROM webhooks WHERE webhook_id = ? AND org_id = ?",
        (webhook_id, user.org_id),
    ).fetchone()
    if not row:
        raise HTTPException(status_code=404, detail="Webhook not found")

    new_secret = secrets.token_hex(32)
    db.conn.execute(
        "UPDATE webhooks SET secret = ? WHERE webhook_id = ? AND org_id = ?",
        (new_secret, webhook_id, user.org_id),
    )
    db.conn.commit()

    return {"webhook_id": webhook_id, "secret": new_secret, "rotated": True}

File: control-plane/src/routes/webhooks.ts
/**
 * Webhooks router — CRUD + test delivery + delivery history + replay + secret rotation.
 */
import { createRoute, z } from "@hono/zod-openapi";
import { createOpenAPIRouter } from "../lib/openapi";
import { ErrorSchema, errorResponses, WebhookCreateBody } from "../schemas/openapi";
import { withOrgDb, withAdminDb } from "../db/client";
import { requireScope } from "../middleware/auth";
import { parseJsonColumn } from "../lib/parse-json-column";

export const webhookRoutes = createOpenAPIRouter();

function genId(): string {
  const arr = new Uint8Array(6);
  crypto.getRandomValues(arr);
  return [...arr].map((b) => b.toString(16).padStart(2, "0")).join("");
}

function validateCallbackUrl(url: string): string | null {
  try {
    const parsed = new URL(url);
    if (!["http:", "https:"].includes(parsed.protocol)) return "Invalid webhook URL";
    const host = parsed.hostname;
    if (!host) return "Invalid webhook URL";
    if (host === "localhost" || host.endsWith(".local") || host.endsWith(".internal")) {
      return "Webhook URL host is not allowed";
    }
    // Basic private IP check
    if (/^(10\.|172\.(1[6-9]|2\d|3[01])\.|192\.168\.|127\.)/.test(host)) {
      return "Webhook URL host is not allowed";
    }
    return null;
  } catch {
    return "Invalid webhook URL";
  }
}

// ── GET /webhooks ───────────────────────────────────────────────────────

const listWebhooksRoute = createRoute({
  method: "get",
  path: "/",
  tags: ["Webhooks"],
  summary: "List all webhooks",
  middleware: [requireScope("webhooks:read")],
  responses: {
    200: {
      description: "List of webhooks",
      content: { "application/json": { schema: z.array(z.record(z.unknown())) } },
    },
  },
});
webhookRoutes.openapi(listWebhooksRoute, async (c): Promise<any> => {
  const user = c.get("user");
  return await withOrgDb(c.env, user.org_id, async (sql) => {
    const rows = await sql`
      SELECT * FROM webhooks ORDER BY created_at DESC
    `;
    return c.json(
      rows.map((r: any) => ({
        webhook_id: r.webhook_id,
        url: r.url,
        events: (() => { try { return JSON.parse(r.events); } catch { return []; } })(),
        is_active: Boolean(r.is_active),
        failure_count: Number(r.failure_count || 0),
        last_triggered_at: r.last_triggered_at || null,
      })),
    );
  });
});

// ── POST /webhooks ──────────────────────────────────────────────────────

const createWebhookRoute = createRoute({
  method: "post",
  path: "/",
  tags: ["Webhooks"],
  summary: "Create a webhook",
  middleware: [requireScope("webhooks:write")],
  request: {
    body: {
      content: {
        "application/json": { schema: WebhookCreateBody },
      },
    },
  },
  responses: {
    200: {
      description: "Webhook created",
      content: { "application/json": { schema: z.record(z.unknown()) } },
    },
    ...errorResponses(400),
  },
});
webhookRoutes.openapi(createWebhookRoute, async (c): Promise<any> => {
  const user = c.get("user");
  const body = c.req.valid("json");
  const url = String(body.url || "").trim();
  const events = Array.isArray(body.events) ? body.events : [];
  const codemodeHandlerId = body.codemode_handler_id || null;

  // If no codemode handler, URL is required
  if (!codemodeHandlerId) {
    const urlError = validateCallbackUrl(url);
    if (urlError) return c.json({ error: urlError }, 400);
  }

  const webhookId = genId();
  const secret = genId() + genId();
  const eventsJson = JSON.stringify(events);

  return await withOrgDb(c.env, user.org_id, async (sql) => {
    await sql`
      INSERT INTO webhooks (id, org_id, url, secret, events, codemode_handler_id)
      VALUES (${webhookId}, ${user.org_id}, ${url || ''}, ${secret}, ${eventsJson}, ${codemodeHandlerId})
    `;

    return c.json({ webhook_id: webhookId, url: url || null, events, codemode_handler_id: codemodeHandlerId });
  });
});

// ── PUT /webhooks/:webhook_id ───────────────────────────────────────────

const updateWebhookRoute = createRoute({
  method: "put",
  path: "/{webhook_id}",
  tags: ["Webhooks"],
  summary: "Update a webhook",
  middleware: [requireScope("webhooks:write")],
  request: {
    params: z.object({ webhook_id: z.string() }),
    body: {
      content: {
        "application/json": {
          schema: z.object({
            url: z.string().optional(),
            events: z.array(z.string()).optional(),
            is_active: z.boolean().optional(),
          }),
        },
      },
    },
  },
  responses: {
    200: {
      description: "Webhook updated",
      content: { "application/json": { schema: z.record(z.unknown()) } },
    },
    ...errorResponses(400),
  },
});
webhookRoutes.openapi(updateWebhookRoute, async (c): Promise<any> => {
  const user = c.get("user");
  const { webhook_id: webhookId } = c.req.valid("param");
  const body = c.req.valid("json");
  const url = String(body.url || "").trim();
  const events = body.events;
  const isActive = body.is_active;

  return await withOrgDb(c.env, user.org_id, async (sql) => {
    if (url) {
      const urlError = validateCallbackUrl(url);
      if (urlError) return c.json({ error: urlError }, 400);
      await sql`UPDATE webhooks SET url = ${url} WHERE webhook_id = ${webhookId}`;
    }
    if (events !== undefined) {
      const eventsJson = JSON.stringify(events);
      await sql`UPDATE webhooks SET events = ${eventsJson} WHERE webhook_id = ${webhookId}`;
    }
    if (isActive !== undefined) {
      await sql`UPDATE webhooks SET is_active = ${isActive ? 1 : 0} WHERE webhook_id = ${webhookId}`;
    }

    return c.json({ updated: webhookId });
  });
});

// ── DELETE /webhooks/:webhook_id ────────────────────────────────────────

const deleteWebhookRoute = createRoute({
  method: "delete",
  path: "/{webhook_id}",
  tags: ["Webhooks"],
  summary: "Delete a webhook",
  middleware: [requireScope("webhooks:write")],
  request: {
    params: z.object({ webhook_id: z.string() }),
  },
  responses: {
    200: {
      description: "Webhook deleted",
      content: { "application/json": { schema: z.record(z.unknown()) } },
    },
    ...errorResponses(404),
  },
});
webhookRoutes.openapi(deleteWebhookRoute, async (c): Promise<any> => {
  const user = c.get("user");
  const { webhook_id: webhookId } = c.req.valid("param");
  return await withOrgDb(c.env, user.org_id, async (sql) => {
    const result = await sql`
      DELETE FROM webhooks WHERE webhook_id = ${webhookId}
    `;
    if (result.count === 0) return c.json({ error: "Webhook not found" }, 404);
    return c.json({ deleted: webhookId });
  });
});

// ── POST /webhooks/:webhook_id/test ─────────────────────────────────────

const testWebhookRoute = createRoute({
  method: "post",
  path: "/{webhook_id}/test",
  tags: ["Webhooks"],
  summary: "Send a test delivery to a webhook",
  middleware: [requireScope("webhooks:write")],
  request: {
    params: z.object({ webhook_id: z.string() }),
  },
  responses: {
    200: {
      description: "Test delivery result",
      content: { "application/json": { schema: z.record(z.unknown()) } },
    },
    ...errorResponses(404),
  },
});
webhookRoutes.openapi(testWebhookRoute, async (c): Promise<any> => {
  const user = c.get("user");
  const { webhook_id: webhookId } = c.req.valid("param");
  return await withOrgDb(c.env, user.org_id, async (sql) => {
    const rows = await sql`
      SELECT * FROM webhooks WHERE webhook_id = ${webhookId}
    `;
    if (rows.length === 0) return c.json({ error: "Webhook not found" }, 404);
    const webhook = rows[0] as any;

    const payload = {
      event: "test",
      timestamp: Date.now() / 1000,
      data: { message: "This is a test webhook delivery from OneShots" },
    };

    try {
      const start = performance.now();
      const resp = await fetch(webhook.url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "X-AgentOS-Secret": webhook.secret,
        },
        body: JSON.stringify(payload),
      });
      const duration = performance.now() - start;
      const respText = await resp.text().catch(() => "");

      // Log delivery (webhook_deliveries is NOT RLS — relies on FK to webhooks)
      await sql`
        INSERT INTO webhook_deliveries (webhook_id, event_type, payload, status_code, response_body, duration_ms, success)
        VALUES (${webhookId}, 'test', ${JSON.stringify(payload)}, ${resp.status}, ${respText.slice(0, 500)}, ${duration}, ${resp.status < 400})
      `;

      return c.json({ status: resp.status, duration_ms: Math.round(duration * 10) / 10, success: resp.status < 400 });
    } catch (e: any) {
      return c.json({ status: 0, error: e.message, success: false });
    }
  });
});

// ── GET /webhooks/:webhook_id/deliveries ────────────────────────────────

const listDeliveriesRoute = createRoute({
  method: "get",
  path: "/{webhook_id}/deliveries",
  tags: ["Webhooks"],
  summary: "List delivery history for a webhook",
  middleware: [requireScope("webhooks:read")],
  request: {
    params: z.object({ webhook_id: z.string() }),
    query: z.object({
      limit: z.coerce.number().int().min(1).max(200).default(50).optional(),
    }),
  },
  responses: {
    200: {
      description: "Delivery history",
      content: { "application/json": { schema: z.record(z.unknown()) } },
    },
    ...errorResponses(404),
  },
});
webhookRoutes.openapi(listDeliveriesRoute, async (c): Promise<any> => {
  const user = c.get("user");
  const { webhook_id: webhookId } = c.req.valid("param");
  const query = c.req.valid("query");
  const limit = Math.min(200, Math.max(1, Number(query.limit) || 50));
  return await withOrgDb(c.env, user.org_id, async (sql) => {
    // Verify webhook ownership via RLS
    const wh = await sql`
      SELECT webhook_id FROM webhooks WHERE webhook_id = ${webhookId}
    `;
    if (wh.length === 0) return c.json({ error: "Webhook not found" }, 404);

    // webhook_deliveries is NOT RLS — gated above by the webhooks RLS check
    const rows = await sql`
      SELECT * FROM webhook_deliveries WHERE webhook_id = ${webhookId}
      ORDER BY created_at DESC LIMIT ${limit}
    `;
    return c.json({ deliveries: rows });
  });
});

// ── POST /webhooks/:webhook_id/deliveries/:delivery_id/replay ───────────

const replayDeliveryRoute = createRoute({
  method: "post",
  path: "/{webhook_id}/deliveries/{delivery_id}/replay",
  tags: ["Webhooks"],
  summary: "Replay a webhook delivery",
  middleware: [requireScope("webhooks:write")],
  request: {
    params: z.object({
      webhook_id: z.string(),
      delivery_id: z.string(),
    }),
  },
  responses: {
    200: {
      description: "Replay result",
      content: { "application/json": { schema: z.record(z.unknown()) } },
    },
    ...errorResponses(404),
  },
});
webhookRoutes.openapi(replayDeliveryRoute, async (c): Promise<any> => {
  const user = c.get("user");
  const { webhook_id: webhookId, delivery_id: deliveryId } = c.req.valid("param");
  return await withOrgDb(c.env, user.org_id, async (sql) => {
    const wh = await sql`
      SELECT * FROM webhooks WHERE webhook_id = ${webhookId}
    `;
    if (wh.length === 0) return c.json({ error: "Webhook not found" }, 404);
    const webhook = wh[0] as any;

    const deliveries = await sql`
      SELECT * FROM webhook_deliveries WHERE id = ${deliveryId} AND webhook_id = ${webhookId}
    `;
    if (deliveries.length === 0) return c.json({ error: "Delivery not found" }, 404);
    const delivery = deliveries[0] as any;

    let payload: any;
    payload = parseJsonColumn(delivery.payload);

    try {
      const start = performance.now();
      const resp = await fetch(webhook.url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "X-AgentOS-Secret": webhook.secret,
        },
        body: JSON.stringify(payload),
      });
      const duration = performance.now() - start;
      const respText = await resp.text().catch(() => "");

      await sql`
        INSERT INTO webhook_deliveries (webhook_id, event_type, payload, status_code, response_body, duration_ms, success)
        VALUES (${webhookId}, ${delivery.event_type || "replay"}, ${JSON.stringify(payload)}, ${resp.status}, ${respText.slice(0, 500)}, ${duration}, ${resp.status < 400})
      `;

      return c.json({
        replayed: deliveryId,
        status: resp.status,
        duration_ms: Math.round(duration * 10) / 10,
        success: resp.status < 400,
      });
    } catch (e: any) {
      return c.json({ replayed: deliveryId, status: 0, error: e.message, success: false });
    }
  });
});

// ── POST /webhooks/:webhook_id/incoming ─────────────────────────────────

const incomingWebhookRoute = createRoute({
  method: "post",
  path: "/{webhook_id}/incoming",
  tags: ["Webhooks"],
  summary: "Process incoming webhook via codemode handler",
  request: {
    params: z.object({ webhook_id: z.string() }),
  },
  responses: {
    200: {
      description: "Incoming webhook processed",
      content: { "application/json": { schema: z.record(z.unknown()) } },
    },
    ...errorResponses(400, 401, 404),
  },
});
webhookRoutes.openapi(incomingWebhookRoute, async (c): Promise<any> => {
  const { webhook_id: webhookId } = c.req.valid("param");

  // Incoming webhook is unauthenticated — look up the webhook + secret via
  // admin connection so we can find ANY org's webhook by id, then verify
  // secret before doing any side effects.
  const webhook = await withAdminDb(c.env, async (sql) => {
    const rows = await sql`
      SELECT * FROM webhooks WHERE webhook_id = ${webhookId} AND is_active = true
    `;
    return rows.length > 0 ? (rows[0] as any) : null;
  });

  if (!webhook) return c.json({ error: "Webhook not found or inactive" }, 404);

  // Verify secret
  const providedSecret = c.req.header("X-AgentOS-Secret") || c.req.header("x-agentos-secret") || "";
  if (webhook.secret && providedSecret !== webhook.secret) {
    return c.json({ error: "Invalid webhook secret" }, 401);
  }

  const payload = await c.req.json().catch(() => ({}));
  const webhookOrgId = String(webhook.org_id || "");

  // If webhook has a codemode handler, process via codemode
  if (webhook.codemode_handler_id) {
    try {
      // Collect incoming headers
      const headers: Record<string, string> = {};
      c.req.raw.headers.forEach((v: string, k: string) => { headers[k] = v; });

      const resp = await c.env.RUNTIME.fetch(
        new Request("https://runtime/api/v1/codemode/webhook-handler", {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            ...(c.env.SERVICE_TOKEN ? { Authorization: `Bearer ${c.env.SERVICE_TOKEN}` } : {}),
          },
          body: JSON.stringify({
            snippet_id: webhook.codemode_handler_id,
            org_id: webhookOrgId,
            payload,
            headers,
          }),
        }),
      );
      const result = await resp.json() as Record<string, unknown>;

      // Log delivery under the webhook owner's org RLS context
      if (webhookOrgId) {
        await withOrgDb(c.env, webhookOrgId, (sql) => sql`
          INSERT INTO webhook_deliveries (webhook_id, event_type, payload, status_code, response_body, duration_ms, success)
          VALUES (${webhookId}, 'codemode', ${JSON.stringify(payload)}, ${resp.status}, ${JSON.stringify(result).slice(0, 500)}, ${0}, ${Boolean(result.processed)})
        `).catch(() => {});
      }

      return c.json(result);
    } catch (err) {
      console.error("[webhook] codemode handler failed:", err);
      return c.json({ error: "Webhook codemode handler failed" }, 502);
    }
  }

  // No codemode handler — forward to webhook URL as before
  if (!webhook.url) return c.json({ error: "Webhook has no URL or codemode handler" }, 400);

  try {
    const start = performance.now();
    const resp = await fetch(webhook.url, {
      method: "POST",
      headers: { "Content-Type": "application/json", "X-AgentOS-Secret": webhook.secret },
      body: JSON.stringify(payload),
    });
    const duration = performance.now() - start;
    const respText = await resp.text().catch(() => "");

    if (webhookOrgId) {
      await withOrgDb(c.env, webhookOrgId, (sql) => sql`
        INSERT INTO webhook_deliveries (webhook_id, event_type, payload, status_code, response_body, duration_ms, success)
        VALUES (${webhookId}, 'incoming', ${JSON.stringify(payload)}, ${resp.status}, ${respText.slice(0, 500)}, ${duration}, ${resp.status < 400})
      `).catch(() => {});
    }

    return c.json({ status: resp.status, duration_ms: Math.round(duration * 10) / 10, success: resp.status < 400 });
  } catch (e: any) {
    return c.json({ status: 0, error: e.message, success: false });
  }
});

// ── POST /webhooks/:webhook_id/rotate-secret ────────────────────────────

const rotateSecretRoute = createRoute({
  method: "post",
  path: "/{webhook_id}/rotate-secret",
  tags: ["Webhooks"],
  summary: "Rotate webhook secret",
  middleware: [requireScope("webhooks:write")],
  request: {
    params: z.object({ webhook_id: z.string() }),
  },
  responses: {
    200: {
      description: "Secret rotated",
      content: { "application/json": { schema: z.record(z.unknown()) } },
    },
    ...errorResponses(404),
  },
});
webhookRoutes.openapi(rotateSecretRoute, async (c): Promise<any> => {
  const user = c.get("user");
  const { webhook_id: webhookId } = c.req.valid("param");
  return await withOrgDb(c.env, user.org_id, async (sql) => {
    const rows = await sql`
      SELECT webhook_id FROM webhooks WHERE webhook_id = ${webhookId}
    `;
    if (rows.length === 0) return c.json({ error: "Webhook not found" }, 404);

    const arr = new Uint8Array(32);
    crypto.getRandomValues(arr);
    const newSecret = [...arr].map((b) => b.toString(16).padStart(2, "0")).join("");

    await sql`
      UPDATE webhooks SET secret = ${newSecret} WHERE webhook_id = ${webhookId}
    `;

    return c.json({ webhook_id: webhookId, secret: newSecret, rotated: true });
  });
});
